internal/exec/exec.go

```diff
@@ -0,0 +1,514 @@
+				handles = append(handles, file.Fd())
+			}
+		}
+
+		// Set up the process to only inherit stdio handles and nothing else.
+		err := e.attrList.Update(
+			windows.PROC_THREAD_ATTRIBUTE_HANDLE_LIST,
+			unsafe.Pointer(&handles[0]),
+			uintptr(len(handles))*unsafe.Sizeof(handles[0]),
+		)
+		if err != nil {
+			return err
+		}
+
+		// Assign the handles to the startupinfos stdio fields.
+		if e.stdioProcEnd[0] != nil {
+			siEx.StdInput = windows.Handle(e.stdioProcEnd[0].Fd())
+		}
+		if e.stdioProcEnd[1] != nil {
+			siEx.StdOutput = windows.Handle(e.stdioProcEnd[1].Fd())
+		}
+		if e.stdioProcEnd[2] != nil {
+			siEx.StdErr = windows.Handle(e.stdioProcEnd[2].Fd())
+		}
+	}
+
+	if e.job != nil {
+		if err := e.job.UpdateProcThreadAttribute(e.attrList); err != nil {
+			return err
+		}
+	}
+
+	if e.cpty != nil {
+		if err := e.cpty.UpdateProcThreadAttribute(e.attrList); err != nil {
+			return err
+		}
+	}
+
+	var zeroSec windows.SecurityAttributes
+	pSec := &windows.SecurityAttributes{Length: uint32(unsafe.Sizeof(zeroSec)), InheritHandle: 1}
+	tSec := &windows.SecurityAttributes{Length: uint32(unsafe.Sizeof(zeroSec)), InheritHandle: 1}
+
+	siEx.ProcThreadAttributeList = e.attrList.List()
+	siEx.Cb = uint32(unsafe.Sizeof(*siEx))
+	if e.execConfig.token != 0 {
+		err = windows.CreateProcessAsUser(
+			e.execConfig.token,
+			argv0p,
+			argvp,
+			pSec,
+			tSec,
+			inheritHandles,
+			flags,
+			createEnvBlock(addCriticalEnv(dedupEnvCase(true, e.env))),
+			dirp,
+			&siEx.StartupInfo,
+			pi,
+		)
+	} else {
+		err = windows.CreateProcess(
+			argv0p,
+			argvp,
+			pSec,
+			tSec,
+			inheritHandles,
+			flags,
+			createEnvBlock(addCriticalEnv(dedupEnvCase(true, e.env))),
+			dirp,
+			&siEx.StartupInfo,
+			pi,
+		)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create process: %w", err)
+	}
+	// Don't need the thread handle for anything.
+	defer windows.CloseHandle(windows.Handle(pi.Thread)) //nolint:errcheck
+
+	// Grab an *os.Process to avoid reinventing the wheel here. The stdlib has great logic around waiting, exit code status/cleanup after a
+	// process has been launched.
+	e.process, err = os.FindProcess(int(pi.ProcessId))
+	if err != nil {
+		// If we can't find the process via os.FindProcess, terminate the process as that's what we rely on for all further operations on the
+		// exec object.
+		if tErr := windows.TerminateProcess(pi.Process, 1); tErr != nil {
+			return fmt.Errorf("failed to terminate process after process not found: %w", tErr)
+		}
+		return fmt.Errorf("failed to find process after starting: %w", err)
+	}
+	return nil
+}
+
+// Run will run the process to completion. This can be accomplished manually by calling Start + Wait afterwards.
+func (e *Exec) Run() error {
+	if err := e.Start(); err != nil {
+		return err
+	}
+	return e.Wait()
+}
+
+// Close will release resources tied to the process (stdio etc.)
+func (e *Exec) Close() error {
+	if e.procState == nil {
```

Comment:
does this need a lock?

Reply:
Going to answer once for all the three lock questions but don't believe so. The idea here is that this object shouldn't be used from multiple threads or have any of the methods called more than once really. os/exec.Cmd has no locking on values it uses to determine whether it can proceed inside certain methods as well. I should add a comment regarding this

https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/os/exec/exec.go;l=499-511

Reply:
Added a comment on the Exec object clarifying some things.